### PR TITLE
Loss functions

### PR DESCRIPTION
**Pull request by hugues31**
_Keras style API_
Implement Loss system
Implement Softmax activation function
Implement Categorical Entropy

For some reason, the MNIST model does not converge.
Regression examples (perceptron and wine quality) do NOT work anymore since the only loss function implemented is the Categorical Entropy. 

**--> We must implement MSE loss function before merging this PR**

### CODE REVIEW

**Review by blhelias — COMMENTED** [review]
A - MA - ZING

File: src/layers/tanh.rs [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    }
+
+    fn backward(&mut self, input: &Tensor, grad_output: Tensor) -> Tensor {
+        let tanh_grad = input.map(|x| Activation::tanh_prime(x));
+        println!("grad tanh: \n{}", grad_output.mult_el(&tanh_grad));
+        grad_output.mult_el(&tanh_grad)
+    }
+}
+
+struct Activation;
```
Comment (hugues31) [inline src/layers/tanh.rs]: Why a new struct ? You can put these 2 functions inside the Tanh struct :smile: and `impl Tanh` instead.

File: src/layers/tanh.rs [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+        format!("Tanh Layer")
+    }
+
+    fn forward(&self, input: &Tensor) -> Tensor {
+        input.map(|x| Activation::tanh(x))
+    }
+
+    fn backward(&mut self, input: &Tensor, grad_output: Tensor) -> Tensor {
+        let tanh_grad = input.map(|x| Activation::tanh_prime(x));
+        println!("grad tanh: \n{}", grad_output.mult_el(&tanh_grad));
```
Comment (hugues31) [inline src/layers/tanh.rs:17]: You can remove it, I suppose it was for testing purpose

File: tests/relu.rs [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+        let layer = ReLU{};
+
+        // Test 1 dimension (when batch = 1 sample for example)
+        let input = Tensor::new(vec![-1.0, 3.0, -2.5, 5.0, -4.0, 2.0], vec![1, 6]);
+
+        let forward = layer.forward(&input);
+
+        let result = Tensor::new(vec![0.0, 3.0, 0.0, 5.0, 0.0, 2.0], vec![1, 6]);
+
+        assert_eq!(utils::round_vector(forward.data, 3), result.data);
```
Comment (hugues31) [inline tests/relu.rs:19]: `round_vector` to 1 decimal maybe instead of 3 ?

File: tests/relu.rs [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+        // Test 1 dimension (when batch = 1 sample for example)
+        let input = Tensor::new(vec![-1.0, 3.0, -2.5, 5.0, -4.0, 2.0], vec![1, 6]);
+
+        let backward = layer.backward(&input, layer.forward(&input));
+        
+        // [0.0, 3.0, 0.0, 5.0, 0.0, 2.0] dot [0.0, 1.0, 0.0, 1.0, 0.0, 1.0]
+        let result = Tensor::new(vec![0.0, 3.0, 0.0, 5.0, 0.0, 2.0], vec![1, 6]);
+
+        assert_eq!(utils::round_vector(backward.data, 3), result.data);
```
Comment (hugues31) [inline tests/relu.rs:34]: Same thing here : 3 -> 1

**Review by hugues31 — COMMENTED** [review]
Well done :+1:

File: src/layers/tanh.rs (hunk 1 above)
Comment (blhelias) [reply src/layers/tanh.rs]: My first attempt was to put every function in `impl Tanh`, but I couldn't find any way to call `tanh` and `tanh_prime` from `forward` and `backward` methods...

Comment (hugues31) [reply src/layers/tanh.rs]: Solved by : e062462
